### DISCUSSION TITLE
Sleep?

### DISCUSSION THREAD

**Discussion by kehrazy** [pytest-dev/pytest#11196] [Q&A] [answered]
The framework is really amazing, but I've been stuck lately.

Say, I have a test:

```python
import time
import pytest

from my_module import my_fixture


def test_my_fixture(my_fixture):
    my_fixture.start_the_test_sequence()
    time.sleep(100)
    # gather results
```

I'm testing hardware, which means I need to wait for hardware, power sources, network adapters etc. etc. When this code gets to `time.sleep` it blocks the fixture, as it should.

I would like for it to not block my fixture. Any help?

**Comment by The-Compiler** [comment]
Can you elaborate on what you mean with "not block the fixture"? What should happen instead? How exactly do you want it to sleep (i.e. wait), yet apparently not do that?

**Reply by kehrazy** [reply]
the flowchart i had in my head goes something along the lines of this
```text
main_thread
│
├── setting up fixtures ── this does something in the background
│                          (say, a server, which has a `.manage()` function)
│
├── sleeping
│
│── (after sleeping, my main thread continues operating)
│
...
│
fixture dies, going on to the next step
```

Fixture is setting up properly. When the control flow gets to `sleeping`, the whole ordeal blocks up. (I am launching it in a threading.Thread).

**Reply by The-Compiler** [reply]
pytest in general is not thread-safe, so this sounds like pretty much uncharted territory you're getting yourself into there anyways. What's the end goal you're trying to accomplish with this?

**Reply by kehrazy** [reply]
> so this sounds like pretty much uncharted territory you're getting yourself into there anyways

yeah, it probably is.

> What's the end goal you're trying to accomplish with this?

like i said, I need to test some hardware. say, I put my eeny-teeny pcb into a furnace, and I wish to see how it performs under pressure.

I need to:
- setup the hardware
- fire up the testing ordeal on the hardware
- read a lot of data from the hardware with python
- manage the hardware

The logic doesn't really matter here, the main point is that it shall be thread-able. Guess it's not possible currently. Maybe i'll make a pull request, if I figure somethin out.

**Reply by The-Compiler** [reply, marked as answer]
I still fail to see how a multi-threaded architecture and an asynchronous sleep come into play. Many of my customers (I give pytest company trainings) test hardware using pytest, and so far I haven't seen a need for this. If you help me understand your use-case, I might be able to give you a better recommendation (perhaps you're looking for something like [pytest-asyncio](https://github.com/pytest-dev/pytest-asyncio)?).

As for making pytest thread-safe, I'd highly recommend opening an issue first and discussing on whether it's a desirable goal at all. Personally, I think the additional complexity costs by far outweigh the (very limited) use-cases it might be needed for.
